fix 'rmi -f imgID' problem

x1022as opened the pull request:
If an image has been tagged to multiple repos and tags, 'docker rmi -f imgID' will just untag one repo other than untagging all repos and deleting the image. This patch fix it.
before:
![before](https://cloud.githubusercontent.com/assets/2840248/6411690/e7e3792e-beb5-11e4-8c74-65842f6740ca.png)
after:
![after](https://cloud.githubusercontent.com/assets/2840248/6411710/2c819926-beb6-11e4-9150-ab0b1561f182.png)

I think 'rmi -f imgID' just untag a random repo:tag is wired. Deleting all repos and tags may be proper

estesp commented on integration-cli/docker_cli_rmi_test.go:
```diff
@@ -77,6 +77,37 @@ func TestRmiTag(t *testing.T) {
+	{
+		imagesAfter, _ := d.Cmd("images", "-a")
+		if strings.Count(imagesAfter, "\n") != strings.Count(imagesBefore, "\n")+4 {
+			t.Fatalf("before: %q\n\nafter: %q\n", imagesBefore, imagesAfter)
```
Can you make this failure a bit clearer to someone who may run the tests but not understand the code yet? Something like "Expected four additional images after tagging; docker images shows: %q".  Not sure the `imagesBefore` is necessary for debugging the test.

GordonTheTurtle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

~~~console
$ git clone -b "upstream-master" git@github.com:x1022as/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

x1022as commented:
@GordonTheTurtle，thanks :)

jfrazelle commented:
this was fixed with https://github.com/docker/docker/pull/11011 but i think we would still like your test can you make sure that rebasing on master still makes your test pass

jfrazelle commented:
ping @coolljt0725 to make sure

estesp commented:
@jfrazelle that PR (#11011) is different in that it only solves the (incorrect) warning a user receives on  an image/tag that isn't going to be removed anyway.. in this PR, the patches are solving the issue that `rm -f` doesn't actually remove all the repos/tags existing for an image.  It would be good to decide if that is the intended behavior, because the current behavior is somewhat random (one tag of the various tags of an image will get removed, but the rest are left around).

jfrazelle commented:
ah ok thnks for clearing it up, the PRs are merging together at this point
in my head

On Mon, Mar 2, 2015 at 11:08 AM, Phil Estes <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle> that PR (#11011
> <https://github.com/docker/docker/pull/11011>) is different in that it
> only solves the (incorrect) warning a user receives on an image/tag that
> isn't going to be removed anyway.. in this PR, the patches are solving the
> issue that rm -f doesn't actually remove all the repos/tags existing for
> an image. It would be good to decide if that is the intended behavior,
> because the current behavior is somewhat random (one tag of the various
> tags of an image will get removed, but the rest are left around).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11058#issuecomment-76784601>.
>

estesp commented:
@x1022as thanks for the test changes; I've reviewed the code and it looks correct, however, I also looked at the docs and there is no clear specification of what `-f` will do with an image tagged in multiple repos or removal of tag by ID with `-f`.

I think you will need to add to this PR an update to the docs (`commandline/cli.md`) with specific details on what `-f` will do, and in some sense this is then a "proposal" of behavior and I want other input from maintainers on whether this is desired result of `-f`.

thaJeztah commented:
> there is no clear specification of what -f will do with an image tagged in multiple repos or removal of tag by ID with -f

Ah, yes. Just remembered I bookmarked this issue for that reason. I hadn't "decided" yet what would be the most wanted outcome of running this command. iirc, some commands will produce an error if *multiple* results (images/containers) are returned; should this do the same (for consistency)?. The fun thing here is of course, should this be regarded *multiple images* or a *single* image, with multiple names/tags?

It could surprise people if they want to remove a specific tag of an image, but use the "id" of the image for convenience and end up removing all aliases of that image.

Perhaps I'm overthinking it, in which case, ignore my rambling

estesp commented:
I think they are mostly good thoughts and why I think this needs broader review as it is more like a proposal and not a simple bug fix.  To your comment:
>It could surprise people if they want to remove a specific tag of an image, but use the "id" of the image for convenience and end up removing all aliases of that image.

I would say that in this case the user has been very silly.. if they are using an ID but wish to only remove a specified tag, they have gone the wrong direction :)  And with the fix from #11011 it should now work properly to remove that single tag without getting a warning about other repo/tags on the same ID.

thaJeztah commented:
> I would say that in this case the user has been very silly

True, happened to do it once, though, to clean up an untagged image, ending up with *another* one being untagged image :smile:

coolljt0725 commented:
This PR is  different from PR( #11011 ) which is fix the misleading/confusing warning when remove a image whose parents has tagged to multiple repository, this PR is to remove a image which has tagged to multiple repository through ID. I have meet this situation when I work PR#11011, at that time I just think it was a desired behavior and didn't realize it was a bug. Since an image could tagged to multiple repository and if I want to remove all the repository of this image, use the image ID to remove all the repository is good way,  or else we have to remove the repository one by one. So I think this PR is good

x1022as commented:
@estesp Thanks for you comment,  I've change this PR according to your advice. please let me know if this PR need any modifications.

x1022as commented:
@jfrazelle when I perform '$ docker rmi IMAGE_ID', docker shows "Conflict, cannot delete image 4986bf8c1536 because it is tagged in multiple repositories, use -f to force"
So I thought ‘$docker rmi -f IMAGE_ID’ will complete this to untag all the repos and tags  and delete the image.
while -f just untag a random repo, which is wired. I thought this should be fixed

icecrime commented:
Ping @dmcgowan @jlhawn @stevvooe: I'm not sure if this is desirable, or has potential interactions with #11109? Thanks!

stevvooe commented on integration-cli/docker_cli_rmi_test.go:
```diff
@@ -77,6 +77,37 @@ func TestRmiTag(t *testing.T) {
+		t.Fatal(err)
+	}
+	imagesBefore, _ := d.Cmd("images", "-a")
+	d.Cmd("tag", "busybox", "utest:tag1")
```
It might be best to check the errors of these commands. Not sure if the results below are viable if these all result in an unchecked error.

x1022as replied:
Thks, I will add error-check

stevvooe replied:
If it fails for another reason that this section is not checking, the failure reason will be lost. In Go, it is good practice to check all error codes, even in test code.

stevvooe commented:
@icecrime The proposed behavior is correct. If one requests the forced removal of an image id, all references to that image id should be removed, as well. Basically, a "-f" accompanied command should do everything in its power to succeed. Anything else would be incorrect if we adhere to a reasonable data model.

Looking at #11109, it's clear that one PR must consider the other.

dmcgowan commented:
I am also +1 for the specified `rmi -f` behavior. For #11109, these should definitely be tested with each other.

x1022as commented:
This pr has little conflict with #11109 on code. and yes, the functionality need to be tested. Comparing the code, it seems that rmi could work fine.

dmcgowan commented:
Rebase needed, now that #11109 is merged I will do another round of testing after rebase.

x1022as commented:
Following code in integration-cli:
```bash
d := NewDaemon(t)
if err := d.StartWithBusybox(); err != nil {
        t.Fatal(err)
} 
```
always failed in windows build, how to deal with this problem?

dmcgowan commented:
Tested with changes from pull by digest, works as expected.  Just waiting for tests to pass

x1022as commented:
Test failed, and I don't know why. Can anyone show me how to fix this?

jfrazelle commented:
It is a registry issue they are looking into it

On Saturday, March 21, 2015, x1022as <notifications@github.com> wrote:

> Test failed, and I don't know why. Can anyone show me how to fix this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11058#issuecomment-84316187>.
>

dmcgowan commented:
LGTM

estesp commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1681,7 +1681,8 @@ deleted.
 an image has one or more tag or digest reference, you must remove all of them
-before the image is removed.
+before the image is removed. If -f and image's short or long ID is specified,
+untag all the names and remove the image.
```
I would suggest this sentence for clarity: "*If the -f flag is used and the image's short or long ID is specified, the command will untag all tags matching the ID and then remove the image itself.*"

estesp commented:
LGTM for code; needs quick docs review--I made a suggestion on clarifying the sentence: @SvenDowideit @moxiegirl @fredlf can make other requests/comments

moxiegirl commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1716,6 +1716,9 @@ You can remove an image using its short or long ID, its tag, or its digest. If
 an image has one or more tag or digest reference, you must remove all of them
 before the image is removed.
 
+If the -f flag is used and the image's short or long ID is specified, the
```
Some constructions are awkward.  This gist contains fix:

https://gist.github.com/6cbfbf7b730ef42ed2c7.git

jamtur01 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1717,6 +1717,9 @@ You can remove an image using its short or long ID, its tag, or its digest. If
 before the image is removed.
 
+If you use the `-f` flag and specify the image's short or long ID, then this
+command untags all tags matching the ID and removes the image itselt.
```
Spelling. itself.

jamtur01 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1739,6 +1742,20 @@ before the image is removed.
     Untagged: test:latest
     Deleted: fd484f19954f4920da7ff372b5067f5b7ddb2fd3830cecd17b96ea9e286ba5b8
 
+Use the `-f` flag to forcibly remove all images at once
```
Add period.

jamtur01 commented:
Docs need some minor edits.

x1022as commented:
@jamtur01 done, thanks.

moxiegirl commented:
LGTM

dmcgowan commented:
Needs rebase

x1022as commented:
I did the rebase and push to my upstream-master branch, but why this PR not updated?

dmcgowan commented:
@x1022as I was having trouble earlier with getting PRs updated, I think github is having difficulties as I was getting a 500 attempting a recreate.  Let's check back tomorrow to see if it is resolved.

x1022as commented:
@dmcgowan thanks, I will check this tomorrow :)

x1022as commented:
Still not updated, how to update this forcibly?

dmcgowan commented:
Try to re-commit and force push again and see if it works. I tried closing and re-opening on another commit and it wouldn't let me re-open.

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1720,6 +1720,9 @@ You can remove an image using its short or long ID, its tag, or its digest. If
 before the image is removed.
 
+If you use the `-f` flag and specify the image's short or long ID, then this
+command untags all tags matching the ID and removes the image itself.
```
I found this very confusing, in part because the example code that follows it is not an example of `-f` usage. As a result, I'm still not quite sure what `-f` does, and I don't know what the example is supposed to be showing me. Below, we have an example of the flag being used to remove images, but this sentence seems to be saying that `-f` is used to remove tags.

I *think* (but am not sure, since I can't actually test the command atm), that this sentence belongs after the one at ln. 1748. Further, we can re-cast the sentence to make it clearer: "The `-f` flag is used to untag and remove all images that match the specified ID."

dmcgowan commented:
@x1022as was there a problem with the rebase? I see a merge commit found its way in

icecrime commented:
@x1022as Some files are not `go fmt`'ed: can you please fix?

```
These files are not properly gofmt'd:
 - daemon/image_delete.go

Please reformat the above files using "gofmt -s -w" and commit the result
```

x1022as commented:
@dmcgowan I did the rebase on another machine, I am not sure if this is the cause. 

I did the push again, and now it works . Thanks

x1022as commented:
@icecrime sorry for that, I'm used to depend on go-fmt tools

estesp commented:
looks like `stripTrailingCharacters` got removed in commit b9d2ede4d4dc67b97abd0254a9e431aa9c89b932

Sorry about that--you need to change to `strings.TrimSpace` as that commit shows. Hopefully this is the last hiccup to get your PR done! :)

x1022as commented:
@estesp Done :)

estesp commented:
I would merge, but I would like @fredlf's comment to be addressed as it seems like he has reasonable concerns about clarity of the doc language: https://github.com/docker/docker/pull/11058#discussion_r27808543

x1022as commented:
I've changed the doc, ping @fredlf to check.

hangyan commented:
Didn't know this issue before,so I opened #12135,they are related,but focused on different situation. I think this issue is about the `-f` flag,we don't have a specified behaviour for it.  In my opinion, it should be restricted to the relationship with containers. Here is all the situations:

1. `docker rmi [-f] <image id> `
    (1)  mapping to one repo and one tag,If no container use it, untag and delete. with containers use this image(not running),`-f` can untag and delete it ,or print error message 
    (2) mapping to one repo and multiple tags,If  no container use,untag all and delete. with containers use  it (not running) ,`-f` can untag all and delete it,or print error message 
    (3) mapping to multiple repo and multiple tags, If no container use ,untag all and delete. with containers use it(not running) ,`-f` can untag all and delete it, or print error message 

#12135 is about `1.(2)`

2. `docker rmi [-f] <name:[tag]>
    (1) this is the  only repo:tag of the image id, same as `1.(1)`
    (2) one repo and many tags mapping to this image, if no container use it, untag all and delete .with containers use it(not running),If no `-f`, remove this tag ,no deletion. with `-f`, untag all and delete
    (3) multiple repo and tags mapping to this image, if no container use it, untag all and delete.with containers use it(not running), if no `-f`, remove this tag,no deletion. with `-f`,untag all and delete

moxiegirl commented:
@jamtur01 ping back

jamtur01 commented:
Docs LGTM

fredlf commented:
LGTM

estesp commented:
Thanks everyone for the reviews!

LGTM

estesp commented:
..but @x1022as it needs a rebase again :( sorry

jfrazelle commented:
so sorry I think we just need a rebase :(

Commit:
'docker rmi -f IMAGE_ID' untag all names and delete the image

If an image has been tagged to multiple repos and tags, 'docker
rmi -f IMAGE_ID' will just untag one random repo instead of
untagging all and deleting the image. This patch implement
this. This commit is composed of:

        *untag all names and delete the image

        *add test to this feature

        *modify commandline/cli.md to explain this

Signed-off-by: Deng Guangxing <dengguangxing@huawei.com>

x1022as commented:
that's ok, rebase done
